torch/csrc/cuda/Module.cpp

```diff
@@ -235,64 +238,118 @@ PyObject * THCPModule_emptyCache(PyObject *_unused, PyObject *noargs)
   Py_RETURN_NONE;
 }
 
-PyObject * THCPModule_memoryAllocated(PyObject *_unused, PyObject *arg)
+PyObject * THCPModule_memoryStats(PyObject *_unused, PyObject *arg)
 {
   HANDLE_TH_ERRORS
   THPUtils_assert(THPUtils_checkLong(arg), "invalid argument to memory_allocated");
-  int device = (int) THPUtils_unpackLong(arg);
-  auto memory_allocated = c10::cuda::CUDACachingAllocator::currentMemoryAllocated(device);
-  return PyLong_FromUnsignedLongLong(memory_allocated);
-  END_HANDLE_TH_ERRORS
-}
+  const int device = (int) THPUtils_unpackLong(arg);
+
+  using c10::cuda::CUDACachingAllocator::StatType;
+  using c10::cuda::CUDACachingAllocator::Stat;
+  using c10::cuda::CUDACachingAllocator::StatArray;
+  using c10::cuda::CUDACachingAllocator::DeviceStats;
+
+  const auto statToDict = [=](const Stat& stat) {
+    PyObject* const dict = THPUtils_newDict();
+    THPUtils_setDictStrInt64(dict, "current", stat.current);
+    THPUtils_setDictStrInt64(dict, "peak", stat.peak);
+    THPUtils_setDictStrInt64(dict, "allocated", stat.allocated);
+    THPUtils_setDictStrInt64(dict, "freed", stat.freed);
+    return dict;
+  };
 
-PyObject * THCPModule_maxMemoryAllocated(PyObject *_unused, PyObject *arg)
-{
-  HANDLE_TH_ERRORS
-  THPUtils_assert(THPUtils_checkLong(arg), "invalid argument to max_memory_allocated");
-  int device = (int) THPUtils_unpackLong(arg);
-  auto max_memory_allocated = c10::cuda::CUDACachingAllocator::maxMemoryAllocated(device);
-  return PyLong_FromUnsignedLongLong(max_memory_allocated);
+  const auto statArrayToDict = [=](const StatArray& statArray) {
+    const std::array<const char*, static_cast<size_t>(StatType::NUM_TYPES)> statTypeNames = {
+      "all", "small_pool", "large_pool"
+    };
+    PyObject* const dict = THPUtils_newDict();
+    for (size_t i = 0; i < statTypeNames.size(); ++i) {
+      THPUtils_setDictStrPyObject(dict, statTypeNames[i], statToDict(statArray[i]));
```

Comment:
I think this leaks references. The documentation for `PyDict_SetItem` doesn't mention that it steals the input reference. That means that you have to decref the original reference after you set it here.

This would all be a lot safer if this code was rewritten in pybind11; I don't really look forward to having to audit all of the other refcounting here.

Reply:
It looks like it doesn't steal the reference, based on (https://docs.python.org/3/c-api/dict.html#c.PyDict_Next). I'll look into fixing this.

Reply:
Separate question - are there any checks in CI for leaky refs? Seems potentially valuable to have if not.

Reply:
Unfortunately stock Python isn't even valgrind clean, you need to stick a bunch of suppressions on it. It might be possible to do some sort of check on memory usage (but it would be very delicate); better to just make it statically impossible to leak refs by using safer APIs.